sklearn/neighbors/base.py

```diff
@@ -300,11 +303,27 @@ class from an array representing our data set and ask who's
         if self._fit_method is None:
             raise ValueError("must fit neighbors before querying")
 
-        X = check_array(X, accept_sparse='csr')
-
         if n_neighbors is None:
             n_neighbors = self.n_neighbors
 
+        if X is not None:
+            query_is_train = False
+            X = check_array(X, accept_sparse='csr')
+        else:
+            query_is_train = True
+            X = self._fit_X
+            # Include an extra neighbor to account for the sample itself being
+            # returned, which is removed later
+            n_neighbors += 1
+
+        train_size = self._fit_X.shape[0]
+        if n_neighbors > train_size:
+            raise ValueError(
+                "Expected n_neighbors <= %d. Got %d" % (train_size, n_neighbors)
+            )
+        n_samples, _ = X.shape
+        sample_range = np.arange(n_samples)[:, None]
```

Comment:
You moved this away from the locality in which it was used. You can move it back now...